server/src/main/java/org/elasticsearch/cluster/coordination/PublicationTransportHandler.java

```diff
@@ -314,96 +342,98 @@ public void onFailure(Exception e) {
         public void sendApplyCommit(DiscoveryNode destination, ApplyCommitRequest applyCommitRequest,
                                     ActionListener<TransportResponse.Empty> listener) {
             assert transportService.getThreadPool().getThreadContext().isSystemContext();
-            transportService.sendRequest(destination, COMMIT_STATE_ACTION_NAME, applyCommitRequest, STATE_REQUEST_OPTIONS,
-                new TransportResponseHandler<TransportResponse.Empty>() {
-
-                    @Override
-                    public TransportResponse.Empty read(StreamInput in) {
-                        return TransportResponse.Empty.INSTANCE;
-                    }
-
-                    @Override
-                    public void handleResponse(TransportResponse.Empty response) {
-                        listener.onResponse(response);
-                    }
-
-                    @Override
-                    public void handleException(TransportException exp) {
-                        listener.onFailure(exp);
-                    }
-
-                    @Override
-                    public String executor() {
-                        return ThreadPool.Names.GENERIC;
-                    }
-                });
+            transportService.sendRequest(
+                destination,
+                COMMIT_STATE_ACTION_NAME,
+                applyCommitRequest,
+                STATE_REQUEST_OPTIONS,
+                new ActionListenerResponseHandler<>(listener, in -> TransportResponse.Empty.INSTANCE, ThreadPool.Names.GENERIC));
         }
 
         private void sendFullClusterState(DiscoveryNode destination, ActionListener<PublishWithJoinResponse> listener) {
-            BytesReference bytes = serializedStates.get(destination.getVersion());
+            assert refCount() > 0;
+            ReleasableBytesReference bytes = serializedStates.get(destination.getVersion());
             if (bytes == null) {
                 try {
                     bytes = serializeFullClusterState(newState, destination.getVersion());
-                    serializedStates.put(destination.getVersion(), bytes);
+                    final ReleasableBytesReference existingBytes = serializedStates.putIfAbsent(destination.getVersion(), bytes);
```

Comment:
I wonder, can't we just do `computeIfAbsent` here instead and block on concurrent access? This is all on the generic pool anyways if there's contention isn't it? In any case no matter the thread this runs on, seems like blocking on the locked CHM key is still cheaper on the system overall than serializing twice? (it's an edge case either way but using `computeIfAbsent` is less code here I think? :)

Reply:
Yeah that's a good point, I wasn't sure if `computeIfAbsent` ran the supplier twice sometimes but you made me go and read the docs and they say that it won't so it should be fine. Definitely a corner case of a corner case to be blocking here but I suppose we may as well not burn CPU while we wait given that we'll be waiting about as long either way. No risk of deadlock either.

Reply:
Ok see 13bfe0c58214c22ec6df00e7d4fba7dc6138b017 (it triggered a few more changes/simplifications too)